### PR TITLE
feat: Remove 'comptime or separate crate' restriction on comptime code

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves <none, this is a bonus>

## Summary\*

Removes the restriction that comptime code can only call into other comptime code or code in a separate crate. This is possible now that functions can be lazily resolved in the interpreter.

This means we no longer need hacky solutions like "comptime impl" on trait impls. `comptime` itself also gets simplified from its previous two meanings of "run this at comptime" and "make this visible to other comptime code" to just "run this at comptime."

## Additional Context

I've also snuck in a sneaky ordering change where we run annotations on traits before annotations on structs. Function annotations are still run last.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [x] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/elaborator/mod.rs [hunk 1]
```diff
@@ -1676,6 +1597,36 @@ impl<'context> Elaborator<'context> {
         }
     }
 
+    /// Run all the attributes on each item. The ordering is unspecified to users but currently
+    /// must run trait attributes first to (e.g.) register derive handlers before derive is
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/mod.rs:1601]:
```suggestion
    /// we run trait attributes first to (e.g.) register derive handlers before derive is
```

**Comment by asterite** [comment]
Should `is_comptime` be gone from `TypeImpl` too?

**Comment by jfecher** [comment]
@asterite yes, updated

**Review by asterite — APPROVED** [review]
Looks good!

**Pull request merged**